Non-XSIMD implementation of Biot Savart kernel for PowerPC

mbkumar opened the pull request:
This PR adds non-XSIMD Biot Savart kernel. 

Present status: 
- [x] Working non-XSIMD code. 
         XSIMD is enabled only for x86_64. I disabled XSIMD for ARM64 and the code is passing tests on my Mac M1 laptop. 
         Note: To test, compile locally and then run tests on the M1/M2 chip. We don't have M1 runners on GitHub yet.
- [x] Compiled the code on PowerPC
        Successful on Traverse @ Princeton. C++ code compiles and simsopt can be installed after removing jax as dependency. Manually compiled older version of jax. Minor edits to some python files to use older jax API.
- [x] Compile VMEC200

mbkumar commented:
I couldn't get OpenMP to be on par with XSIMD. To get speedup with OpenMP, some data structures need to be redesigned, which would touch significant parts of simsopt—modifying the scope of PR to focus on PowerPC only.

andrewgiuliani commented:
Bharat, could you add some timing comparison's to the PR or even just mention the factor slow down you observed here for reference?

mbkumar commented:
@andrewgiuliani 

I am giving the numbers for **Mac M1**

## BiotSavart with XSIMD:
|      N    |     Time (in ms) | Gigainteractions/s | cycles/interaction | 
|-------|----------------|--------------------| ------------------|
| Number of derivatives: 0  | | | |  
  | 1000000         |     1.24    |        0.80645          | 0.029866  |
 | 100000000       |   123      |      0.81301           | 0.029579 |
Number of derivatives: 1
 | 1000000    |          3.06      |       0.3268   |        0.073586 |  
 | 100000000  |        305      |      0.32787  |         0.073264 |
Number of derivatives: 2
| 1000000           |   9.98  |          0.1002   |          0.23955 |
| 100000000       |   990  |          0.10101  |          0.23773 |

## BiotSavartVJP with XSIMD:

|         **N**        | **Time (in ms)** | **Gigainteractions/s** |  **cycles/interaction** |
|------------|------------|--------------------|--------------------|
| Number of derivatives: 0 | | | |
| 1000000          |    2.95     |       0.33898        |   0.070941 |
| 100000000      |    295     |       0.33898        |   0.071006 |
Number of derivatives: 1
| 1000000           |  10.12     |      0.098814         |   0.24296
| 100000000      |   1006     |      0.099404        |    0.24148


## BiotSavart with No SIMD:
|         N        |  Time (in ms) | Gigainteractions/s | cycles/interaction |
|------|----|------|----------|
| Number of derivatives: 0 | | | |
| 1000000    |          1.55      |      0.64516         |  0.037254 |
| 100000000 |         155      |      0.64516         |  0.037367 |
Number of derivatives: 1
| 1000000      |       5.5          |  0.18182       |     0.13223 |
| 100000000  |        551       |     0.18149    |        0.13229 |
Number of derivatives: 2
| 1000000      |       16.29        |   0.061387         |   0.39105 |
| 100000000  |       1625         |  0.061538         |   0.39008 |

## BiotSavartVJP with No SIMD:
|         N        |  Time (in ms) | Gigainteractions/s | cycles/interaction |
|------------|------------|--------------------|--------------------|
Number of derivatives: 0
| 1000000      |        2.95    |        0.33898   |        0.070866 |
| 100000000 |          299   |         0.33445  |         0.071862 |
Number of derivatives: 1
| 1000000       |     17.93    |       0.055772   |         0.43042 |
| 100000000  |       1782   |        0.056117   |         0.42779 |

landreman reviewed (commented):
Is there a way to test the non-xsimd C++ code in the CI? For instance, there could be a compile flag to force use of the non-xsimd implementations even `#if __x86_64__ || __aarch64__`, and apply that flag on some of the runs in the extensive CI.

mbkumar commented on src/simsoptpp/biot_savart_impl.h:
```diff
@@ -3,22 +3,25 @@
 #endif
-
+*/
+#if __x86_64__  || __aarch64__
```
Done

mbkumar commented on src/profiling/profiling.cpp:
```diff
@@ -1,20 +1,101 @@
+
+using namespace std;
+
+#ifdef __x86_64__
```
Done

mbkumar commented:
> Is there a way to test the non-xsimd C++ code in the CI? For instance, there could be a compile flag to force use of the non-xsimd implementations even `#if __x86_64__ || __aarch64__`, and apply that flag on some of the runs in the extensive CI.

I implemented an approach that uses environment variable. If NO_XSIMD environment variable is defined, XSIMD is turned off.

landreman reviewed (commented):
It would be nice if there was a read-only variable available in python so users can check whether xsimd was being used under the hood, maybe a boolean `simsopt.USE_XSIMD`?

mbkumar commented:
Will this work?
``` bash
>>> import simsopt
>>> simsopt.__built_with_xsimd__
False
```

landreman commented on src/simsoptpp/biot_savart_impl.h:
```diff
@@ -284,3 +538,142 @@ void biot_savart_kernel_A(AlignedPaddedVec& pointsx, AlignedPaddedVec& pointsy,
+            auto norm_diff_3_inv = norm_diff_inv*norm_diff_inv*norm_diff_inv;
+
+            auto dgamma_by_dphi_j_simd = Vec3dStd(dgamma_j_by_dphi_ptr[3*j+0], dgamma_j_by_dphi_ptr[3*j+1], dgamma_j_by_dphi_ptr[3*j+2]);
+            /*
+            A_i.x += dgamma_by_dphi_j_simd.x * norm_diff_inv;
+            A_i.y += dgamma_by_dphi_j_simd.y * norm_diff_inv;
+            A_i.z += dgamma_by_dphi_j_simd.z * norm_diff_inv;
+            */
```
Throughout the modified files there are some bits of commented-out code, such as here, which might be good to remove before merging.

landreman commented on src/simsoptpp/biot_savart_vjp_c.h:
```diff
@@ -1,5 +1,14 @@
 #pragma once
 #include "simdhelpers.h"
 
+#if __x86_64__ || __aarch64__
```
`#if defined(USE_XSIMD)`

landreman reviewed (dismissed):
Looks good, thanks. I left one comment about `biot_savart_vjp_c.h` line 4, then it might be good to make a pass through the edited files to remove any commented-out code that is no longer needed. Then ok by me to merge when you're ready.

mbkumar commented:
@landreman 
The fmt library is not compiling on PowerPC anymore. The culprit is PR #306. Let me investigate this before merging the PR.

andrewgiuliani commented:
> > Is there a way to test the non-xsimd C++ code in the CI? For instance, there could be a compile flag to force use of the non-xsimd implementations even `#if __x86_64__ || __aarch64__`, and apply that flag on some of the runs in the extensive CI.
> 
> I implemented an approach that uses environment variable. If NO_XSIMD environment variable is defined, XSIMD is turned off.

Hi Bharat, could this point be added somewhere to the documentation? for example, it could be noted here:

https://simsopt.readthedocs.io/en/latest/cpp.html

andrewgiuliani commented on src/simsoptpp/vec3dsimd.h:
```diff
@@ -177,24 +200,371 @@ inline Vec3dSimd cross(int i, Vec3dSimd& b){
+    return a.x * a.x + a.y * a.y + a.z * a.z;
+}
+
+struct alignas(ALIGN_BYTES) Vec3dSimdPortable1 {
```
Admittedly, Vec3d was not documented previously, but now that there are two structures in use:
Vec3dstd and Vec3dPortable1, I think for readability the purpose of these two structures should be documented.  Why do we need both?

mbkumar replied:
I removed Vec3dPortable1 since it is not used in production. And I added comments for Vec3dStd.

andrewgiuliani commented on src/simsoptpp/biot_savart_impl.h:
```diff
@@ -161,10 +163,254 @@ void biot_savart_kernel(AlignedPaddedVec& pointsx, AlignedPaddedVec& pointsy, Al
+#else
+
+template<class T, int derivs>
+void biot_savart_kernel_openmp(AlignedPaddedVec& pointsx, AlignedPaddedVec& pointsy, AlignedPaddedVec& pointsz,
```
Is this only called in the profiling code?

mbkumar replied:
Yes. Since this is not used in production code, I deleted this.

mbkumar commented:
I fixed the issue with the compilation of fmt library by making it header only and preventing it from being compiled by updating the CMakeLists.txt. simsopt once again compiles on PowerPC and the tests are passing.

I also removed Vec3dSimdPortable1 class and the BiotSavart kernel that depends on Vec3dSimdPortable1. 
 
This PR is ready to be merged.

mbkumar commented on src/simsoptpp/biot_savart_vjp_impl.h:
```diff
@@ -172,9 +175,112 @@ void biot_savart_vjp_kernel(AlignedPaddedVec& pointsx, AlignedPaddedVec& pointsy
+          throw std::runtime_error("res_grad_dgamma_by_dphi needs to be in row-major storage order");
+    int num_points         = pointsx.size();
+    int num_quad_points    = gamma.shape(0);
+    constexpr int simd_size = 1;
```
In the NON-XSIMD code, simd_size=1 implies no vectorization. Since I am using constexpr, it will be optimized away during compilation. Anyhow I deleted the variable now, but it shouldn't matter due to compile-time optimization of constexpr variables.
